Only throttle failed login attempts

walterl opened the pull request:
This will not work until metabase/throttle#1 has been merged, and Metabase's deps updated to include that change.

walterl committed:
Compute Content-Security-Policy hashes for inline JS (#10504)

* Split out inline JS from index/init templates to separate files

* Read inline JS from resources at run time

* Calculate inline JS hashes for CSP header from content

* Move inline JS to resource sub-directory

* Update and memoize inline JS loading

* Revert debug code

* Deduplicate `resp/response` calls

* Fix paths to moved inline JS resources

* Force creation of test data DB so things don't get left in the cache

This fix was provided by @camsaul.

* Combine and `defonce` inline JS hashes

* s/inlinejs/inline-js/

walterl committed:
Ensure that Metabase JAR is readable by `metabase` container user (#10510)

If the umask of the user that builds the image is xx7, the JAR will not be
world readable in the resulting image.

walterl committed:
Remove `--add-modules=java.xml.bind` Java option (#10521)

Fixes #10244. @camsaul confirmed that it is no longer needed.

walterl committed:
Join on collection table's ID, not the joining table's `collection_id` (#10481)

* Join on collection table's ID, not the joining table's `collection_id`

This caused each record that joins on collection, to be duplicated for each
existing collection:

`LEFT JOIN collection collection ON card.collection_id = collection_id`

should be

`LEFT JOIN collection collection ON card.collection_id = collection.id`

* Test fixes from @cam on EE

sbelak committed:
Insights: correctly handle infinities in results (#10502)

Filter out models with infinities

walterl committed:
Transfer log events as JSON objects and allow filtering on front-end (#10522)

* Return log events as JSON objects from API; render on front-end

Site UUID added to log events.

* Remove superfluous memoization of site UUID retrieval

Settings are already cached.

* Transfer timestamps in ISO-8601 format

* Move `metabase.metabot.instance/local-process-uuid` to `metabase.public-settings`

* Use `local-process-uuid` in stead of `site-uuid`

`site-uuid` is not unique across processes, which is what we want to
associate with log events.

* Remove unused import

* Allow log events to be filtered by process UUID on front-end

* Move process selector out of loading wrapper

Otherwise, if there are process UUIDs with no log events, the selector is
removed with the log event list.

* Figured out how to add docs to `defonce`

* Fix ns declaration

* Linter appeasement

* Merge log events received from back-end with previously received ones

This allows responses from multiple back-end instances to be spliced
together.

* Use MB components in stead of plain, styled HTML tags

The select is only displayed if more than one process UUID was found.

* Linter appeasement

* Move `metabase.public-settings/local-process-uuid` to `metabase.config`

* Fix sorting on multiple log event fields

* Flatten `Select` children to allow nested arrays of children

* Generate options directly under parent Select

* Limit the number of log events that we render

* Docstring and style updates

* Update editor config with Clojure files max line length

* Ensure that `NumericWrapper` is imported before use

* Cleanup layout

walterl committed:
Bump HoneySQL version (#10559)

The new version (0.9.5) includes @camsaul's Turkish upper case fix:
jkk/honeysql#237

walterl committed:
Factor out test for access to localStorage (#10533)

* Factor out test for access to localStorage

This allows code that uses localStorage to test against
`window.hasLocalStorage`, in stead of handling exceptions generated when
access to localStorage is denied.

* Move check for localStorage to metabase/lib/dom

Commit:
Fix typo in warning message (#10556)

kdoh committed:
revert text color to #74838f (#10560)

* revert text-medium color to #74838f

* use light instead of medium on preview button

sbelak committed:
MBTL (#10105)

Add transforms and domain models

victoriaspek committed:
Added page about maps and updated nav (#10495)

* Added page about maps and updated nav

* Clarified step 1

* Added more details throughout

* Quick URL change

* Added additional resource!

* fix typo: 'sdisplay' to 'display'

paulrosenzweig committed:
Add message with link to docs for fields without any filter widgets (#10340)

salsakran committed:
Correct proxy command line [ci skip] (#10430)

walterl committed:
Bump Toucan version to include Turkish lower case fix (#10581)

walterl committed:
Wrap `tru` and `trs` macro results in `str` (#10571)

* Add versions of `trs` and `tru` that wrap results in `str`

The pre-existing `trs` and `tru` were renamed to `lazy-trs` and `lazy-tru`
respectively, and new `trs` and `tru` funcs were added, wrapping the result
of its lazy counterpart in `str`.

This way the normal use case of `trs` and `tru` is improved by no longer
requiring that it be wrapped in `str`.

Cases where the translated result is passed to `str` anyway, the lazy
version can be used.

* `(str (trs ...))` → `(trs ...)`

And the same for `tru`.

* Allow `defsetting` descriptions to use lazy versions of `trs`/`tru`

* Use `lazy-trs`/`lazy-tru` where `str` is applied separately

* Remove unnecessary `vec` from `tru`/`trs` macros

* puppetlabs.i18n does not have a `lazy-tru`

* Missed some settings that now need to use `lazy-tru`

* Only use `lazy-tru` in top-level `def`s

If the localized strings are looked up on module import, before locales are
loaded, the lookup will fail and the original input string will be used.

* Fix bad usages of `tru` that should be `lazy-tru`

* s/lazy-tr/deferred-tr/g

* Avoid `tr*` macros being used in top-level (compile time) statements

Because localizations are not yet loaded at compile time.

* Add type hint for `str*` wrapping of `trs`/`tru`

* Fix assigned function's return type hint

* Expand docstrings for `i18n/tr[su]`

* Use `trs` and `tru` from metabase.util.i18n instead of puppetlabs.i18n.core

* Defer compile time translation lookups

These changes were pointed out by exceptions generated by `i18n/str*`.

* Remove unused `require`d function

* New compile time `trs` usages caught by `lein uberjar`

* Ensure that error message from db driver is a string

It is necessary to ensure that the error message is a string, because some
db driver errors come from statically declared error messages, using
`deferred-tru`. Like those in
`metabase.driver.common/connection-error-messages`.

Fixes #10347 and #10473.

* Remove string type hint from `humanize-connect-error-message`

walterl committed:
Only throttle failed login attempts

This will not work until metabase/throttle#1 has been merged, and
Metabase's deps updated to include that change.

sbelak committed:
Don't disable preview for fields with non-trivial type  (#9687)

walterl committed:
Don't create API callback if no database is associated with card (#10573)

* Don't create API callback if no database is associated with card

Fixes #10538

* Return empty array to avoid mapping over null

walterl committed:
Throttle logins on request support, with custom header support

If the `:mb-session-throttle-source-header` config option is set, a login
request's source is read from that header in the HTTP request, instead of
using `(:remote-addr request)`. The concrete use case for this is setting
is `X-Forwarded-For`.

walterl committed:
Update test expectations to reflect fixes in `throttle` lib

walterl committed:
Expand `X-Forwarded-For` header source IP test

flamber commented:
Hi @walterl 
A question if I may.

Reading your commit comment, it seems like the throttle functionality won't be working for everyone who is using a reverse proxy unless they add the following variable, correct?
(And of course send that proxy header, like Nginx's `proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;`)

`MB_SESSION_THROTTLE_SOURCE_HEADER="X-Forwarded-For"`

Would it be possible to possible to use the specified header if variable exists, but otherwise automatically try to read `X-Forwarded-For` if exist and fallback to `Remote-Addr` ?
Or would it be too dangerous to change throttle IP without the admin make that change manually?

walterl commented:
Hi @flamber. Yes, it will [fall back](https://github.com/metabase/metabase/pull/10593/commits/945b8e6219f1781af6a9461638c2dbb9504b17b9#diff-4749fc04c2fa4d280c75a64dd5ecb54bR107) to using the remote address if no header was specified in the `MB_SESSION_THROTTLE_SOURCE_HEADER` option, or that header has no value.

This is specifically to support reverse proxy type cases. Even without specifying a header, it just means that the 50 request threshold for login attempts will be shared by all clients connecting through the proxy.

Since the other major part of this block of work is to only throttle _failed_ login attempts, 50 will probably be sufficient in most cases.

walterl committed:
Add test for per request source throttling

walterl committed:
Rename variable

walterl committed:
Factor out injection of HTTP 400 status into error

walterl committed:
Replace remaining uses of `throttle-check` with `with-throttling`

walterl committed:
Re-add `throttle-check` for use in `forgot_password` endpoint

The usage semantics of this endpoint is different, in that it doesn't have
a concept of the failed request. Therefore all requests need to be
throttled.

victoriaspek committed:
Operations guide breakout (#10569)

* Splitting the operations guide into separate pages

* Added a description to the Operations Guide nav

* Renamed the files and updated the navigation

walterl committed:
Bump throttle version

walterl committed:
Initialize new `source-address-header` setting during setup

walterl committed:
Replace request source header config with public setting

walterl committed:
Update `with-throttling` usage to new multi-throttler form

walterl committed:
Parameterize `cleaned-throttlers` to make it more generally applicable

walterl committed:
Add test for throttling on `forgot_password` endpoint

walterl committed:
Update docstring

walterl committed:
Only throttle Google auth requests when throttling is enabled

walterl committed:
Don't initialize source address header setting last

It nukes the returned response object from `set-session-cookie`.

walterl committed:
Linter fixes and appeasement

camsaul commented on src/metabase/api/setup.clj:
```diff
@@ -93,6 +93,9 @@
     ;; notify that we've got a new user in the system AND that this user logged in
     (events/publish-event! :user-create {:user_id (u/get-id new-user)})
     (events/publish-event! :user-login {:user_id (u/get-id new-user), :session_id (str session-id), :first_login true})
+    ; set `source-address-header` to "X-Forwarded-For" if such a header is present
```
https://guide.clojure.style/#two-semicolons-for-code-fragment

walterl replied:
Grarg, sorry :man_facepalming:

camsaul commented on test/metabase/api/session_test.clj:
```diff
@@ -56,6 +58,73 @@
+            :body
+            json/parse-string
+            (get "errors")
+            (get "username"))
```
Instead of multiple calls to `get` you can use [`get-in`](http://clojuredocs.org/clojure.core/get-in) e.g. `(get-in <whatever> ["errors" "username"])`

camsaul commented on src/metabase/api/session.clj:
```diff
@@ -97,16 +97,46 @@
+    (if-let [header-value ((:headers request) source-address-header)]
+      header-value
+      (:remote-addr request))
+    (:remote-addr request)))
```
You can make this a lot more concise with a bit of destructuring and `or`, e.g.

```clj
(defn- source-address
  [{:keys [headers remote-addr]}]
  (or (some->> (public-settings/source-address-header) (get headers))
      remote-addr))
```

You could even do

```clj
(defn- source-address
  [{{source-addr (public-settings/source-address-header)} :headers, :keys [remote-addr]}]
  (or source-addr remote-addr))
```

but if `source-address-header` is nil and for some insane reason there's a nil value in headers (e.g. `{nil "blah blah blah"}`) you'll get that value back as `source-addr`, so I would go with the former

camsaul commented on src/metabase/api/setup.clj:
```diff
@@ -93,6 +93,9 @@
     (events/publish-event! :user-create {:user_id (u/get-id new-user)})
     (events/publish-event! :user-login {:user_id (u/get-id new-user), :session_id (str session-id), :first_login true})
+    ; set `source-address-header` to "X-Forwarded-For" if such a header is present
+    (when (-> request :headers (get "x-forwarded-for"))
```
I would consider using `get-in` here instead of threading, `(get-in request [:headers "x-forwarded-for"])` is both more concise and IMO clearer, but I don't feel too strongly about this.

camsaul commented on src/metabase/api/setup.clj:
```diff
@@ -93,6 +93,9 @@
     (events/publish-event! :user-login {:user_id (u/get-id new-user), :session_id (str session-id), :first_login true})
+    ; set `source-address-header` to "X-Forwarded-For" if such a header is present
+    (when (-> request :headers (get "x-forwarded-for"))
+      (public-settings/source-address-header "X-Forwarded-For"))
```
If headers always come back from ring lowercased, then this looks like it might not do what you expect the `source-address` function.

If I were you I would give `public-settings/source-address-header` a magic get that lowercases the value (hopefully in an English locale 😉) as it comes out so it doesn't matter if you pass in an uppercase value. The docstring for `defsetting` explains how to do that. That way a user doesn't need to know the value of the setting has to be lowercased if they set it themselves, e.g. via the `MB_SOURCE_ADDRESS_HEADER` env var

camsaul reviewed (changes requested):
*  We need to make sure `source-address-header` is always lowercased because Ring headers are always lowercased, you can use a `defsetting` magic getter to do that

victoriaspek committed:
FAQ section of the docs (#10587)

* FAQ section of the docs

* Made links relative

* Link & TOC changes

* Fixing duplicate header, adding data collection URL

* Removed the "option is turned off by default" wording

walterl committed:
Don't return table data for archive searches (#10627)

Fixes #10439

flamber committed:
Fix docs links, small text changes (#10613)

sbelak committed:
Fingerprinting: correctly handle datetimes that are also FKs (#10621)

camsaul committed:
Merge branch 'release-0.33.x' into master

paulrosenzweig committed:
Center chart with legend (#10406)

walterl committed:
Simplify function implementation

walterl committed:
Replace consecutive `get`s with single `get-in`s

walterl committed:
Fix comment styling

walterl committed:
Ensure that `source-address-header` setting is always (safely) lower cased

walterl committed:
Add test for `metabase.util/lower-case-end`

walterl committed:
Merge branch 'master' into login-throttling-updates-8736

# Conflicts:
#	project.clj

camsaul reviewed (approved):
This looks good 😎 

Can you change the target branch to `release-0.33.x` so we can be sure it gets released as part of 0.33.1?
